File: CourseRegistrationSystem/src/Advisor.java
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.ArrayList;
import java.util.List;
public class Advisor extends User {
    @JsonProperty("advisedStudents")
    private ArrayList<String> advisedStudentIDs;
    private ArrayList<Student> requestedStudents;
    private ArrayList<Student> advisedStudents;

    public Advisor(String username, String name, String surname, String password, ArrayList<String> advisedStudentIDs) {
        super(username, name, surname, password);
        this.advisedStudentIDs = advisedStudentIDs;

    }

    public Advisor(){}

    void setAdvisedStudentsInit(List<Student> students){
        advisedStudents= new ArrayList<>();
        for (Student s: students ){
            if (advisedStudentIDs.stream().anyMatch(p -> p.equals(s.getUsername()))){
                advisedStudents.add(s);
            }
        }
    }

    public  List<Student> listRequestStudents() {
        requestedStudents = new ArrayList<>();
        for(Student st: advisedStudents){
            if (st.getRequestedCourses().size()>0){
                requestedStudents.add(st);
            }
        }

        return requestedStudents;
    }


    public void approveCourseRegistration(Student student, Course course) {

        course.addEnrolledStudent(student,course);

    }

    public String studentsToString() {
        StringBuilder result = new StringBuilder();

        // Headers
        String headers = String.format("%-15s %-40s\n", "Student ID", "Full Name");


        if (!advisedStudents.isEmpty()) {
            result.append("Advised Students:\n");
            result.append(headers);
            for (Student student : advisedStudents) {

                String studentInfo = String.format("%-15s %-40s\n",
                        student.getStudentId(), student.getName() + " " + student.getSurname());
                result.append(studentInfo);
            }
        } else {
            result.append("No advised students.\n");
        }

        return result.toString();
    }
}

File: iteration3/Advisor.py
from typing import List
from User import User
from Student import Student
from Course import Course

class Advisor(User):
    def __init__(self, username="", name="", surname="", password="", advised_student_ids=None):
        super().__init__(username, name, surname, password)
        self.advised_student_ids = advised_student_ids if advised_student_ids else []
        self.requested_students = []
        self.advised_students = []

    def set_advised_students_init(self, students: List['Student']):
        self.advised_students = [s for s in students if s.username in self.advised_student_ids]

    def list_requested_students(self) -> List['Student']:
        self.requested_students = [st for st in self.advised_students if len(st.requested_courses) > 0]
        return self.requested_students

    def approve_course_registration(self, student: 'Student', course: 'Course'):
        course.add_enrolled_student(student, course)

    def students_to_string(self) -> str:
        result = ""
        headers = f"  {'Student ID':<15} {'Full Name':<40}\n"

        if self.advised_students:
            result += "Advised Students:\n"
            result += headers
            for i, student in enumerate(self.advised_students, 1):
                student_info = f"{i}- {student.student_id:<15} {student.name} {student.surname}\n"
                result += student_info
        else:
            result += "No advised students.\n"

        return result

    def get_advised_students(self) -> List['Student']:
        return self.advised_students
